Added Docs for AGP 9 Migration (#12825)
flutter/website: src/content/release/breaking-changes/update-to-AGP-9.md

---

> title: Breaking Changes from Android Gradle Plugin 9.0.0

Reviewer: ```suggestion
title: Migrating Flutter Android app to Android Gradle Plugin 9.0.0
```
We use sentence case, not title case, but I think this title might be a bit clearer.

Author: Done

description: >-
  How to migrate your Flutter app's Android Gradle files 
  to build apps with Android Gradle Plugin 9.0.0+.
---

## Summary

To build a Flutter app for Android, the Android Gradle Plugin (AGP)
must be applied. Starting AGP 9.0.0, a few breaking changes have been made.

> The following migrations must be made in order to successfully apply AGP 9+.

Reviewer: ```suggestion
must be applied. As of AGP 9.0.0, 
the following migrations are required to successfully apply AGP 9+.
```

First, built-in Kotlin is the new default, meaning any apps
using the `kotlin-android` plugin will not build successfully.
You must migrate from `kotlin-android` to built-in Kotlin.

Second, AGP 9+ will only use the new AGP DSL interfaces.
This means any old DSL types will not be properly read. 
The Flutter team is working on migrating old DSL types 
to use the new DSL. In the meantime, you can set a gradle property flag 

> to allow for usage of the old DSL.

Author: Done

In Flutter 3.40.0, support was added for applying AGP 9+. Projects created

> with versions of Flutter prior to 3.40.0 need to be migrated manually.

Author: Done

To learn more about Android Gradle Plugin,

> see [Android Gradle Plugin docs][AGP block].

Author: Done

## Migrate

> These instructions assume you are updating from an AGP version < 9.0.0

Reviewer: ```suggestion
These instructions assume you are updating from an AGP version before 9.0.0
```

Author: done

to an AGP version 9.0.0+.

> You should also be using minimum compatible dependency versions

Author: done

listed in the [Android Gradle Plugin docs][AGP block].

> ### android/app/build.gradle or android/app/build.gradle.kts

Reviewer: ```suggestion
### Update the gradle file
```

> If your app does not apply the `kotlin-android` (Kotlin Gradle Plugin), then

Author: done

skip to the next step.

> First, find the `kotlin-android` plugin likely located in the plugins block of

Author: done

the `<app-src>/android/build.gradle` or `<app-src>/android/build.gradle.kts`
file.
As an example, consider the `build.gradle.kts` file from
a new Flutter app created before this change:

```kotlin
plugins {
    id("com.android.application")
    id("kotlin-android")
    // The Flutter Gradle Plugin must be applied after the Android and Kotlin Gradle plugins.
    id("dev.flutter.flutter-gradle-plugin")
}

android {
    ...
    kotlinOptions {
        jvmTarget = JavaVersion.VERSION_17.toString()
    }
    ...
}
...
```

> Next, remove the `kotlin-android` plugin and the `kotlinOptions` block like so:

Reviewer: ```suggestion
Next, remove the `kotlin-android` plugin and the `kotlinOptions` block:
```

Author: Done

```kotlin diff
plugins {
    id("com.android.application")
    -id("kotlin-android")
    // The Flutter Gradle Plugin must be applied after the Android and Kotlin Gradle plugins.
    id("dev.flutter.flutter-gradle-plugin")
}

android {
    ...
    -kotlinOptions {
        -jvmTarget = JavaVersion.VERSION_17.toString()
    -}
    ...
}
```

Replace the `kotlinOptions` block with the following:

```kotlin diff
+kotlin {
    +compilerOptions {
        +jvmTarget = org.jetbrains.kotlin.gradle.dsl.JvmTarget.JVM_17
    +}
+}
```

Here is how the file will likely end up:

```kotlin
plugins {
    id("com.android.application")
    // The Flutter Gradle Plugin must be applied after the Android and Kotlin Gradle plugins.
    id("dev.flutter.flutter-gradle-plugin")
}

android {
    ...
    kotlin {
        compilerOptions {
            jvmTarget = org.jetbrains.kotlin.gradle.dsl.JvmTarget.JVM_17
        }
    }
    ...
}
...
```

> ### android/gradle.properties

Author: Done

Moving on, to allow usage of the old AGP DSL, set the gradle property flag

> `android.newDsl` to false in your app's `<app-src>/android/gradle.properties`.

Reviewer: ```suggestion
`android.newDsl` to `false` in your app's `<app-src>/android/gradle.properties` file.
```

```properties diff
org.gradle.jvmargs=-Xmx8G -XX:MaxMetaspaceSize=4G -XX:ReservedCodeCacheSize=512m -XX:+HeapDumpOnOutOfMemoryError
android.useAndroidX=true
+ android.newDsl=false
```

> ### Validation

Reviewer: ```suggestion
### Validate
```

Author: Done

Execute `flutter run` to confirm that your app builds and
launches on a connected Android device or emulator.

## Timeline

Support in stable release: 3.40.0

> Recommended in stable release: 3.43.0

Reviewer: What does "recommended in stable release: xxx" mean? Does it mean that they should make these changes by this release?

Author: My mistake. AGP 9 is actually not automatically supported/used/included in newly created Flutter apps yet. It will be in a future release after AGP 9 is officially released to stable after 1/8/26. I updated the timeline section accordingly.

## References

Relevant issue:

> [Issue #175688][]

Author: done

The Gradle build files in your app vary based on the Flutter version 

> used when generating your app. 

Reviewer: ```suggestion
The Gradle build files in your app vary based on the Flutter version 
used when your app was created. 
```

Author: Done

> You should consider staying up-to-date with the latest version 

Author: Done

of the build files by running `flutter upgrade` 

> in your app's directory periodically.

Author: Done

---
title: Breaking Changes from Android Gradle Plugin 9.0.0
description: >-
  How to migrate your Flutter app's Android Gradle files to
  successfully build Flutter apps when upgrading to 
  Android Gradle Plugin 9.0.0+.
---

## Summary

To build a Flutter app for Android, the Android Gradle Plugin (AGP) must be applied.
Starting AGP 9.0.0, a few breaking changes have been made. The following migrations
must be made in order to successfully apply AGP 9+.
First, built-in kotlin is the new default, meaning any apps using the android-kotlin

> plugin will not build successfully. You must migrate from android-kotlin to built-in kotlin.

Author: done

Second, AGP 9+ will only use the new AGP DSL interfaces, meaning any old DSL
types will not be properly read. We are working on migrating old DSL types
to use the new DSL. In the meantime, you can set a gradle property flag
to allow for usage of the old DSL.

In Flutter 3.40.0, support was added for applying AGP 9+. Projects created
with versions of Flutter prior to 3.40.0 need to be migrated manually.

To learn more about Android Gradle Plugin, see [Android Gradle Plugin docs][AGP block].

## Migrate

These instructions assume you are updating from an AGP version < 9.0.0 to an AGP version 9.0.0+ and 
you are using minimum compatible dependency versions listed in the 

> [Android Gradle Plugin docs][AGP block]. 

Author: done

    ndkVersion = flutter.ndkVersion

    compileOptions {
        sourceCompatibility = JavaVersion.VERSION_17
        targetCompatibility = JavaVersion.VERSION_17
    }

    kotlin {
        compilerOptions {
            jvmTarget = org.jetbrains.kotlin.gradle.dsl.JvmTarget.JVM_17
        }
    }

    defaultConfig {
        // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
        applicationId = "com.example.brand_new"
        // You can update the following values to match your application needs.
        // For more information, see: https://flutter.dev/to/review-gradle-config.
        minSdk = flutter.minSdkVersion
        targetSdk = flutter.targetSdkVersion
        versionCode = flutter.versionCode
        versionName = flutter.versionName
    }

    buildTypes {
        release {
            // TODO: Add your own signing config for the release build.
            // Signing with the debug keys for now, so `flutter run --release` works.
            signingConfig = signingConfigs.getByName("debug")
        }
    }
}

flutter {
    source = "../.."
}
```

### android/gradle.properties

> Moving on, to allow usage of the old AGP DSL, set the gradle property flag `android.newDSL` to

Author: done

false in your app's `<app-src>/android/gradle.properties`.

```properties diff
org.gradle.jvmargs=-Xmx8G -XX:MaxMetaspaceSize=4G -XX:ReservedCodeCacheSize=512m -XX:+HeapDumpOnOutOfMemoryError
android.useAndroidX=true
+ android.newDsl=false
```

### Validation

Execute `flutter run` to confirm that your app builds and
launches on a connected Android device or emulator.

## Timeline

Support in stable release: 3.40.0
Recommended in stable release: 3.43.0

## References

Relevant issue:
[issue #175688][]

Gradle build files generated by `flutter create`
differ across Flutter versions.
You should consider using the

> latest versions of build files.

Author: done

        // You can update the following values to match your application needs.
        // For more information, see: https://flutter.dev/to/review-gradle-config.
        minSdk = flutter.minSdkVersion
        targetSdk = flutter.targetSdkVersion
        versionCode = flutter.versionCode
        versionName = flutter.versionName
    }

    buildTypes {
        release {
            // Signing with the debug keys for now, so `flutter run --release` works.
            signingConfig = signingConfigs.getByName("debug")
        }
    }
}

flutter {
    source = "../.."
}
```

### android/gradle.properties

Moving on, to allow usage of the old AGP DSL, set the gradle property flag
`android.newDsl` to false in your app's `<app-src>/android/gradle.properties`.

```properties diff
org.gradle.jvmargs=-Xmx8G -XX:MaxMetaspaceSize=4G -XX:ReservedCodeCacheSize=512m -XX:+HeapDumpOnOutOfMemoryError
android.useAndroidX=true
+ android.newDsl=false
```

### Validation

Execute `flutter run` to confirm that your app builds and
launches on a connected Android device or emulator.

## Timeline

Support in stable release: 3.40.0
Recommended in stable release: 3.43.0

## References

Relevant issue:

> [issue #175688][]

Author: done

Gradle build files generated by `flutter create`
differ across Flutter versions.
You should consider using the latest versions 

> of build files.

Reviewer 2: This comment feels like it should live somewhere else but I'll leave that up to @Sfshaza.
 
Either way, I think you should clarify this note a bit by letting folks know how to use the latest version, e.g.

```md
The Gradle build files in your app vary based on the Flutter version you used when you generated your app. You should consider staying up to date with the latest version of the build files by running `flutter create` in your app's directory periodically.
```
